cassiopeia/solver/SharedMutationJoiningSolver.py

```diff
@@ -0,0 +1,298 @@
+        nb_similarity_function,
+        similarity_map: pd.DataFrame,
+        cherry: Tuple[str, str],
+        new_node: str,
+        missing_state_indicator: int = -1,
+        weights=None,
+    ) -> pd.DataFrame:
+        """Update similarity map after finding a cherry.
+
+        Adds the new LCA node into the character matrix with the mutations
+        shared by the joined nodes as its character vector. Then, updates the
+        similarity matrix by calculating the pairwise similarity between the
+        new LCA node and all existing nodes.
+
+        Args:
+            character_matrix: Contains the character information for all nodes,
+                updated as nodes are joined and new internal LCA nodes are added
+            nb_similarity: A numba compiled similarity function
+            similarity_map: A similarity map to update
+            cherry: A tuple of indices in the similarity map that are joining
+            new_node: New node name, to be added to the updated similarity map
+            missing_state_indicator: Character representing missing data
+            weights: Weighting of each (character, state) pair. Typically a
+                transformation of the priors.
+
+        Returns:
+            A new similarity map, updated with the new node
+        """
+
+        character_i, character_j = (
+            np.where(character_matrix.index == cherry[0])[0][0],
+            np.where(character_matrix.index == cherry[1])[0][0],
+        )
+
+        character_array = character_matrix.to_numpy(copy=True)
+        similarity_array = similarity_map.to_numpy()
+        i_characters = character_array[character_i, :]
+        j_characters = character_array[character_j, :]
+        lca = data_utilities.get_lca_characters(
+            [i_characters, j_characters], missing_state_indicator
+        )
+        character_matrix.loc[new_node] = lca
+
+        similarity_array_updated = self.__update_similarity_map_numba(
+            character_array,
+            similarity_array,
+            np.array(lca),
+            nb_similarity_function,
+            missing_state_indicator,
+            weights,
+        )
+
+        sample_names = list(similarity_map.index) + [new_node]
+
+        similarity_map = pd.DataFrame(
+            similarity_array_updated, index=sample_names, columns=sample_names
+        )
+
+        # drop out cherry from similarity map and character matrix
+        similarity_map.drop(
+            columns=[cherry[0], cherry[1]],
+            index=[cherry[0], cherry[1]],
+            inplace=True,
+        )
+
+        character_matrix.drop(
+            index=[cherry[0], cherry[1]],
+            inplace=True,
+        )
+
+        return similarity_map
+
+    @staticmethod
+    @numba.jit(nopython=True)
+    def __update_similarity_map_numba(
+        character_matrix: np.array,
+        similarity_map: np.array,
+        lca: np.array,
+        nb_similarity_function,
```

Comment:
Missing typing for the `nb_similarity_function`

Reply:
What would be the typing for a numba compiled function out of curiosity?

Reply:
I think you can still call it a `Callable`, right? Does this break the `numba` compilation?